### PR TITLE
chore: updated ESLint to v9 with flat config

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## Description

Bumps ESLint to v9 and migrates the ESLint config from the legacy "eslintrc" format (`apps/site/.eslintrc.json`) to the new "flat" format (`apps/site/eslint.config.js`). This brought a few complications 🙃...

* Three plugins don't yet have stable flat config support:
  * Next.js is ... not at all? https://github.com/vercel/next.js/discussions/49337
  * React is in canary: https://github.com/facebook/react/issues/28313
  * Storybook is in canary: https://github.com/storybookjs/eslint-plugin-storybook/issues/135#issuecomment-2182186572
* `import` doesn't have ESLint v9 support (https://github.com/import-js/eslint-plugin-import/issues/2948), so I moved to its successor/rival/something-like-that [`import-x`](https://github.com/un-ts/eslint-plugin-import-x)

This also butted into some pre-existing issues:

* `stylelint-selector-bem-pattern` unintentionally has a dependency on an old version of `eslint-plugin-jest`, which has a dependency on an old version of `@typescript-eslint/utils`: https://github.com/simonsmith/stylelint-selector-bem-pattern/issues/72.
  * So, even though this PR uses the latest typescript-eslint version, there's still the annoying _`WARNING: You are currently running a version of TypeScript ...`_ (https://github.com/nodejs/nodejs.org/pull/6991#discussion_r1746981252)
* ESLint's `--cache` doesn't work properly when you use cross-file linting, such as `import`'s cross-fine analysis. I've removed it altogether in this PR. See: https://typescript-eslint.io/troubleshooting/faqs/eslint#can-i-use-eslints---cache-with-typescript-eslint

The lines of ESLint config code are about the same _if_ we ignore temporary flat config compatibility layers and disables for transient issues:

* `main`: ~120 (~100 `.eslintrc.json` + ~20 `.eslintignore`)
* `eslint-9`: ~120 (~150 `eslint.config.js` - ~25 compat lines - ~5 transient disables)

## Validation

Broadly, `npm run lint`. Specifically, `npm run lint:js` and `npm run lint:md` from `apps/site`.

I also verified performance isn't worse in this PR. It's actually a little better! Using `time npm run` + `lint:js` or `lint:md`, ignoring Turbo caches, on an M1 Max Mac Studio...

<table>
<thead>
<tr>
<th>Files</th>
<th><code>main</code></th>
<th><code>eslint-9</code></th>
</tr>
</thead>
<tbody>
<tr>
<td><code>lint:js</code></td>
<td>~10s</td>
<td>~5s</td>
</tr>
<tr>
<td><code>lint:md</code></td>
<td>~55s</td>
<td>~55s</td>
</tr>
</tbody>
</table>

<table>

## Related Issues

Parallel to #6991.

### Check List

- [x] I have read the [Contributing Guidelines](https://github.com/nodejs/nodejs.org/blob/main/CONTRIBUTING.md) and made commit messages that follow the guideline.
- [x] I have run `npm run format` to ensure the code follows the style guide.
- [x] I have run `npm run test` to check if all tests are passing.
- [x] I have run `npx turbo build` to check if the website builds without errors.
- [x] I've covered new added functionality with unit tests if necessary.

### CODE REVIEW

File: apps/site/util/hexToRGBA.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-export const hexToRGBA = (hex: string = '', alpha = 0.9) => {
+export const hexToRGBA = (hex = '', alpha = 0.9) => {
```
Comment (JoshuaKGoldberg) [inline apps/site/util/hexToRGBA.ts:1]: Not strictly necessary for this PR, but I figured I'd throw it in 😄. This was a violation of [`@typescript-eslint/no-inferrable-types`](https://typescript-eslint.io/rules/no-inferrable-types). That's enabled in the _stylistic_ config, which isn't enabled in `main` or this PR.

File: apps/site/layouts/Download.tsx [hunk 2]
```diff
@@ -10,6 +10,7 @@ import styles from './layouts.module.css';
 const DownloadLayout: FC<PropsWithChildren> = async ({ children }) => {
   const {
     frontmatter: { title, subtitle },
+    // eslint-disable-next-line react-hooks/rules-of-hooks
   } = useClientContext();
```
Comment (JoshuaKGoldberg) [inline apps/site/layouts/Download.tsx:14]: Without this disable:

```plaintext
React Hook "useClientContext" cannot be called in an async function.
eslint[react-hooks/rules-of-hooks](https://reactjs.org/docs/hooks-rules.html)
```

I haven't kept up-to-date with the hullabaloo around async functions and hooks. This function doesn't seem to need to be async to me, but 🤷 I don't want to put forward a potentially risky behavior change in this PR.

File: apps/site/components/withChangelogModal.tsx [hunk 3]
```diff
@@ -62,7 +62,7 @@ const WithChangelogModal: FC<WithChangelogModalProps> = ({
             }
           );
         }
-      } catch (_) {
+      } catch {
```
Comment (JoshuaKGoldberg) [inline apps/site/components/withChangelogModal.tsx:65]: Behavior change in ESLint v9 and typescript-eslint v8: catch blocks are now also checked by default in [`no-unused-vars`](https://eslint.org/docs/latest/rules/no-unused-vars`) -> [`@typescript-eslint/no-unused-vars`](https://typescript-eslint.io/rules/no-unused-vars). 🧹 

* https://eslint.org/blog/2024/04/eslint-v9.0.0-released/#updates-to-existing-rules has the full list of breaking rule behavior changes in ESLint v9.
* https://typescript-eslint.io/blog/announcing-typescript-eslint-v8#rule-breaking-changes has the full list of breaking rule behavior changes in typescript-eslint v8.

File: apps/site/components/withDownloadCategories.tsx [hunk 4]
```diff
@@ -13,6 +13,7 @@ const WithDownloadCategories: FC<PropsWithChildren> = async ({ children }) => {
   const t = await getTranslations();
   const releases = await getReleaseData();
 
+  // eslint-disable-next-line react-hooks/rules-of-hooks
   const { pathname } = useClientContext();
```
Comment (JoshuaKGoldberg) [inline apps/site/components/withDownloadCategories.tsx:17]: Without this disable:

```plaintext
React Hook "useClientContext" cannot be called in an async function.
eslint[react-hooks/rules-of-hooks](https://reactjs.org/docs/hooks-rules.html)
```

I haven't kept up-to-date with the hullabaloo around async functions and hooks. 🤷

**Review by bmuenzenmeyer — APPROVED** [review]
Working locally on my machine (Windows 10)
Thanks so much for this work @JoshuaKGoldberg

**Review by AugustinMauroy — APPROVED** [review]
LGTM, thanks for giving time on that !

**Comment by ovflowd** [comment]
With all due respect, can I just hug you, Josh?!! Thank you so much for this contribution! Very very much appreciated 😭

**Review by canerakdas — APPROVED** [review]
LGTM!

**Review by shanpriyan — APPROVED** [review]
LGTM! 🚀

**Pull request merged**